[BEAM-5392] group by key non merging window optimalization

mareksimunek opened the pull request:
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

mareksimunek committed:
[BEAM-5392] GroupByKey optimized for non-merging windows

dmvk commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java:
```diff
@@ -0,0 +1,201 @@
+                  w,
+                  windowingStrategy.getWindowFn().getOutputTime(windowedValue.getTimestamp(), w));
+      WindowedValue<K> keyWinValue =
+          WindowedValue.of(key, timestamp, w, windowedValue.getPane()); //todo predavat pane
```
todo??

dmvk commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java:
```diff
@@ -0,0 +1,201 @@
+
+    @Override
+    public int compare(CompositeKey o1, CompositeKey o2) {
+      int keyCompare = SignedBytes.lexicographicalComparator().compare(o1.getKey(), o2.getKey());
```
`Unsigned`

dmvk commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java:
```diff
@@ -0,0 +1,201 @@
+   * @param <K> type of key iterator emits
+   * @param <V> type of value iterator emits
+   */
+  static class ReduceByKeyIterator<K, V, W extends BoundedWindow>
```
`GroupByKeyIteratoror`

dmvk commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java:
```diff
@@ -0,0 +1,201 @@
+        Iterators.transform(itr, t2 -> t2._1.withValue(KV.of(t2._1().getValue(), t2._2())));
+  }
+
+  public static class CompositeKeyComparator implements Comparator<CompositeKey>, Serializable {
```
private

Commit:
[BEAM-5987] Cache and share materialized side inputs between Spark tasks

mareksimunek committed:
[BEAM-5987] Use Guava's Cache to control eviction of materialized side inputs

brianmartin committed:
[BEAM-6211] Support deflate (zlib) in CompressedFile

`.deflate` files are quite common in Hadoop and also supported by
TensorFlow in TFRecord file format. Moreover, `.deflate` is already
supported since 0.6.0 by the Java SDK (see BEAM-1518).

robertwb committed:
[BEAM-6243] Add an experiment to use Python's optimizer on Flink.

jklukas committed:
[BEAM-6365] Add ZStandard compression support for Java SDK

Adds a dependency on `zstd-jni` which declares no transitive dependencies,
but otherwise just fills out enum values to expose the ZStandard support
that already exists in commons-compress.

Commit:
[BEAM-6138] Set the PTransform name on Java SDK User counters

RobbeSneyders committed:
Partially port avroio

markflyhigh committed:
[BEAM-5091] Wordcount integration test on DataflowRunner in Python 3

ryan-williams committed:
add toString()s to Metrics classes

- MetricKey
- MetricName
- MetricQueryResults
- MetricsContainerStepMap

ryan-williams committed:
add MetricResult.result wrapper around committed || attempted

ryan-williams committed:
fix MetricQueryResults comments

RobbeSneyders committed:
Skip failing tests

kennknowles committed:
Use virtualenv to build Python release zip

markflyhigh committed:
Sync dependencies in base_image_requirements.txt with setup.py

adude3141 committed:
add post build action to record errorpron/checkstyle/findbugs on precommit java

lgajowy committed:
Merge pull request #7612: [BEAM-6207] Added option to publish synthetic data to Kafka topic

mxm committed:
[BEAM-4520] Add note on checkpointing of UnboundedSources

This prints a warning when checkpointing is disabled and pipelines are run with
the FlinkRunner which contain UnboundedSources. Most UnboundedSources rely on
checkpoint acknowledgement, e.g. PubSubIO.

mxm committed:
[BEAM-4520] Warn when UnboundedSources are used without checkpointing

This prints a warning when checkpointing is disabled and pipelines are run with
the FlinkRunner which contain UnboundedSources. Most UnboundedSources rely on
checkpoint acknowledgement, e.g. PubSubIO.

lgajowy committed:
[BEAM-6463] Fix DAG shape for load tests (branched where it shouldn't)

lgajowy committed:
[BEAM-6463] Quickfix NPE when there's no optional stepOptions parameter

chamikaramj committed:
Merge pull request #7367: [BEAM-3342] Create a Cloud Bigtable Python connector Write

tvalentyn committed:
Update minimal version of futures due to transitive requirement in google-api-core 1.7.0

kennknowles committed:
Merge pull request #7667: [BEAM-6545] Remove useless allowance of null constructor in ByteArrayShufflePosition

Commit:
[BEAM-6431] Implement Execution Time metrics start,process,finish in the Java SDK

kennknowles committed:
Roll our own more strict nullness annotation

kennknowles committed:
[BEAM-5658] Re-enable findbugs in Dataflow worker

kennknowles committed:
[BEAM-5658] Fixes return of null where it is forbidden

chamikaramj committed:
Upgrades gcsio dependency to 1.9.12

kennknowles committed:
[BEAM-5658] Fixes null warning in StreamingDataflowWorker

kennknowles committed:
[BEAM-5658] Fix misuse of Random in StreamingDataflowWorker

kennknowles committed:
[BEAM-5658] Remove redundant null checks in Dataflow worker

kennknowles committed:
[BEAM-5658] Clarify risky System.exit in Dataflow worker

kennknowles committed:
[BEAM-5658] Fix accidental use of integer division in Dataflow worker GroupingTables

kennknowles committed:
[BEAM-5658] Remove @Nullable param that always has checkNotNull

udim committed:
[BEAM-5959] Add GCS KMS support

GCS operations refactored to support objects with KMS keys. Beam will
not set KMS keys on GCS objects. Users can utilize bucket default keys.

Also adds --dataflowKmsKey, which is passed to DataflowRunner.

Details:
- GCS copy operation reimplemented as rewrite - allowing copies of
objects using KMS keys (source or dest).
  - Rewrite also supports copying across regions and storage classes.
  https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite
- Introduces --dataflowKmsKey flag, which should apply to Dataflow
pipeline state.
  - If creating a new bucket for gcpTempLocation, sets the bucket
  default key to --dataflowKmsKey.
- New integration tests:
  - GcsKmsKeyIT - pipeline with GCS sink using --dataflowKmsKey and
  gcpTempLocation.
  - GcsUtilIT - tests multi-part rewrite against prod GCS

aaltay committed:
Merge pull request #7652 from markflyhigh/py3-simple-wc

[BEAM-5091] New wordcount integration test on DataflowRunner in Python 3

chamikaramj committed:
Merge pull request #7678: [BEAM-6552] Upgrades gcsio dependency to 1.9.12

kennknowles committed:
[BEAM-5658] Suppress findbugs warning about intended reference equality check

kennknowles committed:
[BEAM-5658] Use %n instead of \n in Dataflow worker

kennknowles committed:
[BEAM-5658] Suppress non-serializable warnings due to superclasses

kennknowles committed:
[BEAM-5658] Clarify unreachability of null worker state in Dataflow worker

kennknowles committed:
Use checkArgument instead of checkNotNull for checking argument

kennknowles committed:
[BEAM-5658] Make some finalize methods protected

kennknowles committed:
[BEAM-5658] Make mutable field merely package-public, not public

kennknowles committed:
[BEAM-5658] Remove dead field from WindmillStateInternals

kennknowles committed:
[BEAM-5658] Eliminate needless boxing and unboxing

kennknowles committed:
[BEAM-5658] Make constant field static

kennknowles committed:
[BEAM-5658] Separate various null checks to make more legible

kennknowles committed:
[BEAM-5658] Suppress probable synchronization bugs in Dataflow worker, with Jiras filed

kennknowles committed:
[BEAM-5658] Suppress nullability error in IsmReaderImpl to be fixed when we can

kennknowles committed:
[BEAM-5658] Class starts thread in constructor - make it final so this is OK

kennknowles committed:
[BEAM-5658] note that GroupingShuffleRangeTracker is OK with null groupStart

kennknowles committed:
[BEAM-5658] remove noop equals override from ExecutionStateTracker

kennknowles committed:
[BEAM-5658] Suppress and note many synchronization errors in Dataflow worker

kennknowles committed:
[BEAM-5658] Fix erroneous cases of integer division in Dataflow worker

kennknowles committed:
[BEAM-5658] Fix nullability in Dataflow worker AvroByteReader and factory

kennknowles committed:
[BEAM-5658] Fix minor code issues in Dataflow worker

kennknowles committed:
[BEAM-5658] Fix nullability checking in Dataflow worker

kennknowles committed:
[BEAM-5658] Suppress warning about unread field in Dataflow worker DebugCapture

kennknowles committed:
Remove spammy println from BeamModulePlugin.applyPortabilityNature

kennknowles committed:
Suppress warning caused by our upstream dependencies

robertwb committed:
Rename compact_accumulator to compact for consistancy.

Also resolve two related TODOs.

aromanenko-dev committed:
Merge pull request #7436: Update design-your-pipeline.md

mxm committed:
Merge pull request #7673: [BEAM-4520] Warn when UnboundedSources are used without checkpointing

Juta committed:
[BEAM-5315] Python 3 port bigquery and pubsub related io modules

mxm committed:
Merge pull request #7656: [BEAM-4775] Metrics improvements: make wordcount.py work with attempted metrics

robertwb committed:
Merge pull request #7661 [BEAM-4030] Add CombineFn.compact for Python.

robertwb committed:
Register another generic iterable type.

lgajowy committed:
Merge pull request #7674: [BEAM-6463] Load test pipelines fixes

robertwb committed:
Avoid taking abs(MIN_INT) which is undefined.

kanterov committed:
[BEAM-6497] Retry ClickHouse container startup

- do 10 retries to fix "container not found" error
- use later docker image to avoid possible problem

robertwb committed:
Merge pull request #7297  [BEAM-6243] Add an experiment to use Python's optimizer on Flink.

kyle-winkelman committed:
[BEAM-4783] Fix issues created in #6181.

kennknowles committed:
Merge pull request #7689: [BEAM-6497] Retry ClickHouse container startup

robertwb committed:
Merge pull request #7686 Register another generic iterable type.

tvalentyn committed:
[BEAM-6569] Remove a flaky test from a precommit suite. (#7691)

* [BEAM-6569] Remove a flaky test from a precommit suite.

* Remove duplicate mentions of apache_beam.io.gcp.gcsfilesystem_test, apache_beam.io.hdfs_integration_test, apache_beam.io.range_trackers_test, and once again apache_beam.io.gcp.gcsio_test.

charlesccychen committed:
Merge pull request #7657 from markflyhigh/py3-update-dill-version

Update dependencies in base_image_requirements.txt to sync with setup.py

kennknowles committed:
[BEAM-5658] Suppress findbugs warning about pointer-based compareTo override

Commit:
Update design-documents.md

charlesccychen committed:
Merge pull request #7669 from tvalentyn/update-futures

Update minimal version of futures due to transitive requirement in google-api-core 1.7.0

kennknowles committed:
Merge pull request #7668: [BEAM-6547] Release automation: use virtualenv; use SSH URL

iemejia committed:
[BEAM-6426] Enforce ban of non-vendored Guava, with exceptions

kennknowles committed:
[BEAM-6574] Inline description of CSV formats

kennknowles committed:
Merge pull request #7684: [BEAM-5658] Re-enable findbugs on Dataflow worker

adude3141 committed:
[BEAM-1132] add jacoco report on javaPreCommit

kennknowles committed:
[BEAM-6566] Add plugin to check IWYU dependencies, inactivated

kennknowles committed:
[BEAM-6566] Activate IWYU for Beam SQL and fix errors

akedin committed:
Merge pull request #7697 from kennknowles/CSVFormat

[BEAM-6574] Inline description of CSV formats

iemejia committed:
Merge pull request #7695: [BEAM-6426] Enforce ban of non-vendored Guava, with exceptions

mxm committed:
Merge pull request #7666: Add post build action to record errorprone/checkstyle/findbugs on precommit java

aromanenko-dev committed:
Merge pull request #7694: Update design-documents.md with link to Get Metrics API

aaltay committed:
Merge pull request #7685 from Juta/io-gcp

[BEAM-5315] Python 3 port bigquery and pubsub related io modules

pabloem committed:
[BEAM-5959] Adding the gcp_kms_key option to Python SDK. (#7696)

* Adding the gcp_kms_key option to Python SDK.

* fix typo

Commit:
spotless

Commit:
change a comment

markflyhigh committed:
[BEAM-5091] Run wordcount_it in Python 3 in ValidatesContainer test

Commit:
added one more test

Commit:
spotless

Commit:
addressed comments

Commit:
spotless

charlesccychen committed:
Python 3 fixes for GCP auth and Avro IO

charlesccychen committed:
Work around issue in dill for wordcount example save_main_session

charlesccychen committed:
Update Dataflow container, select correct Python 3 container

markflyhigh committed:
[INFRA-17335] Check more Python versions in inventory job

swegner committed:
Merge pull request #7700: [BEAM-6558] Add IWYU plugin, activate for Beam SQL, fix errors

RobbeSneyders committed:
Skip vcfio tests

mxm committed:
[BEAM-6019] Report root cause of failure in Flink portable pipelines

The `JobInvocation` only reported the top-most message in the stack trace. This
changes the reporting to pass on the root cause instead.

Unblocks several Python PVR tests which rely on exception reporting.

robertwb committed:
Merge pull request #7688 from robertwb/timestamp-div-fixes

Avoid taking abs(MIN_INT) which is undefined.

tvalentyn committed:
Use Dataflow dev container images that have current versions of Beam SDK dependencies installed.

swegner committed:
Merge pull request #7624: [BEAM-6138] Set the PTransform name on Java SDK User counters

tvalentyn committed:
Regenerate BQ client to match apitools==0.5.26

mxm committed:
Merge pull request #7716: [BEAM-6019] Report root cause of failure in Flink portable pipelines

tvalentyn committed:
Set response encoding for BigQuery APIs to 'utf-8' on Python3

chamikaramj committed:
Merge pull request #7253: [BEAM-6211] Support deflate (zlib) in CompressedFile

Commit:
spotless and others

lostluck committed:
[BEAM-3306] Add first pass coder registry for Go

chamikaramj committed:
Upgrades gcsio dependency to 1.9.13

charlesccychen committed:
Merge pull request #7721 from tvalentyn/fix_bq_client

[BEAM-6532] Regenerate BQ client to match apitools==0.5.26 and fix BQ io to work on Python 3

charlesccychen committed:
Merge pull request #7644 from RobbeSneyders/avroio

[BEAM-5315] Partially port Avroio module

charlesccychen committed:
Merge pull request #7710 from charlesccychen/fix-wordcount-py3

Work around issue in dill for wordcount example save_main_session

charlesccychen committed:
Merge pull request #7708 from charlesccychen/py3-fix

Python 3 fixes for GCP auth and Avro IO

htyleo committed:
[BEAM-6577] Examine func return types when generating shims (#7720)

* Examine func return types when generating shims

starcgenx.Extractor should examine both function input and output when extracting types from a package. Otherwise, the generated code may miss certain package imports.

* Refactor extractParameters

Rename extractParameters to extractFromSignature, which calls extractFromTuple using both params and results.

aaltay committed:
Merge pull request #7718 from tvalentyn/patch-35

Use Dataflow dev container images that have current versions of Beam SDK dependencies installed.

charlesccychen committed:
Merge pull request #7711 from charlesccychen/update-dfpy3-container

Update Dataflow container, select correct Python 3 container

chamikaramj committed:
Merge pull request #7724: [BEAM-6582] Upgrades gcsio dependency to 1.9.13

charlesccychen committed:
Remove EXPERIMENTAL_BEAM_PY3 and add a warning for Python 3 execution

charlesccychen committed:
Fix BigQuery RowAsDictJsonCoder output type in Python 3

aaltay committed:
Merge pull request #7723 from lostluck/userdefinedcoder

[BEAM-3306] Add first pass coder registry for Go

Commit:
spotless

udim committed:
[BEAM-5959] KMS support for BigQuery

Adds withKmsKey() methods to reads and writes. Affects tables created
tables by: streaming writes, batched writes, query temporary tables,
table copies (for very large batched writes).

Commit:
added some test

Commit:
Add more tests and the SimpleStateRegistry

reuvenlax committed:
Merge pull request #7714: [BEAM-6579] Fix create disposition for the case of temporary tables

robertwb committed:
Revert "Revert "Merge pull request #7316 [BEAM-6269] Cross-SDK transform expansion protocol.""

This reverts commit a65f1c43b7fbedfec3e7396d9e024bf0e762c33d.

charlesccychen committed:
Merge pull request #7727 from charlesccychen/fix-bq-coder

Fix BigQuery RowAsDictJsonCoder output type in Python 3

charlesccychen committed:
Merge pull request #7726 from charlesccychen/py3-warning

Remove EXPERIMENTAL_BEAM_PY3 and add a warning for Python 3 execution

robertwb committed:
Use ints rather than strings for language boundary crossing.

robertwb committed:
Revert "Create a standard coder for utf-8 encoded strings."

This reverts commit afe1a8436aafe5203df24ac9f37d15b1fb6ad04c.

robertwb committed:
[BEAM-6587] Make StringUtf8Coder a standard coder in Python.

robertwb committed:
Make grpc dependency optional.

je-ik committed:
[BEAM-6589] include org.objenesis into shaded jar

dmvk committed:
Merge pull request #7732: [BEAM-6589] include org.objenesis into shaded jar

markflyhigh committed:
Py3 IT on DirectRunner in PreCommit

kkucharc committed:
[BEAM-6291] Generic BigQuery schema load tests metrics (#7614)

* [BEAM-6291] Generic schema for BQ load tests

* [BEAM-6291] Added pipeline option to check if metrics are required

* [BEAM-6291] Common environment variable to disable load tests.

* [BEAM-6291] Refactored BQ and Metrics load tests utils. Added missing documentation and pipelineoption.

markflyhigh committed:
fixup! fix scripts directory

kennknowles committed:
Merge pull request #7699: [BEAM-1132] add jacoco report on javaPreCommit

kennknowles committed:
[BEAM-6595] Do not push to release branch from gradle maven release plugin

Commit:
Iterated on PR

charlesccychen committed:
Fix broken link in BigQuery Python connector docs

lostluck committed:
Update starcgen for new types

amaliujia committed:
[BEAM-4522] Re-enable CalcRemoveRule

swegner committed:
Merge pull request #7676: [BEAM-6431] Implement Execution Time metrics start,process,finish in the Java SDK

aaltay committed:
Merge pull request #7706 from markflyhigh/py3-wordcount-it

[BEAM-5091] Run integration test in Python 3 in ValidatesContainer test

robertwb committed:
[BEAM-3742] Checkpointing for SDF over FnAPI in Python SDK and FnApiRunner. (#7722)

charlesccychen committed:
Merge pull request #7734 from markflyhigh/py3-wc-direct

[BEAM-6593] Python 3 IT on DirectRunner in PreCommit

aaltay committed:
Merge pull request #7738 from lostluck/fixshims

Update starcgen for new types

charlesccychen committed:
Merge pull request #7740 from charlesccychen/fix-broken-bq

Fix broken link in BigQuery Python connector docs

drieber committed:
Got rid of reference to windmill_service_streaming_rpc_batching experiment (#7728)

* There is no need to predicate WindmillServiceStreamingRpcBatchLimitFactory on the windmill_service_streaming_rpc_batching experiment. That experiment is now fully rolled out.

* Fixed argument to streamingEngineEnabled.

robertwb committed:
[BEAM-6269] Cross-SDK transform expansion protocol. (#7729)

* Revert "Revert "Merge pull request #7316 [BEAM-6269] Cross-SDK transform expansion protocol.""

This reverts commit a65f1c43b7fbedfec3e7396d9e024bf0e762c33d.

* Use ints rather than strings for language boundary crossing.

* Revert "Create a standard coder for utf-8 encoded strings."

This reverts commit afe1a8436aafe5203df24ac9f37d15b1fb6ad04c.

* Make grpc dependency optional.

aaltay committed:
Fix a bug in use of no_use_multiple_sdk_containers experiment flag.
Setting this flag should prevent the addition of
use_multiple_sdk_containers experiment.

sekikn committed:
[BEAM-6601] Fix wrong command to run Hadoop InputFormat integration test in Javadoc

aaltay committed:
Merge pull request #7741 from aaltay/expb

Fix a bug in use of no_use_multiple_sdk_containers experiment flag.

kennknowles committed:
Merge pull request #7739: [BEAM-4522] Re-enable CalcRemoveRule

iemejia committed:
Merge pull request #7742: [BEAM-6601] Fix wrong command to run Hadoop InputFormat integration test in Javadoc

nielm committed:
Check for null values when encoding Key columns.

When encoding Key Column values, if the column value is
unspecified, assume that the value is null.

mxm committed:
[BEAM-6608] Prepare to-be-staged files early enough in Flink Runner

Before execution, the Flink Runner prepares all necessary files, in order for
them to be shipped to the remote cluster.

Due to a recent regression via BEAM-2873, the preparation is performed after the
execution environment has been initialized. This skips the preparation phase and
results in invalid files being shipped.

mxm committed:
[BEAM-6609] Default tempLocation in FlinkPipelineOptions to default tmp directory

There should be a sensible default, so users don't have to configure it by default.

mxm committed:
Fix temp path having to end with a slash in PipelineResources

lostluck committed:
Use the same noescape that the Go standard library in ioutilx

lostluck committed:
Remove noisy logging.

aaltay committed:
Merge pull request #7749 from lostluck/typos

Use the same noescape that the Go standard library in ioutilx

aaltay committed:
Merge pull request #7750 from lostluck/warnings

Remove noisy logging.

dustin12 committed:
[BEAM-6540] Windmill quota piping (#7664)

* [BEAM-6540] Report Throttling due to Streaming RPC Quota

iemejia committed:
[BEAM-6426] Fix regexp to enforce ban of non-vendored Guava

Fix the matiching uses of non-vendored guava and adds suppressions for
cases when guava is part of the components API.

udim committed:
Website: Update built-in IO table

kennknowles committed:
[BEAM-6613] Make svn commit of source release less interactive, more idempotent

kennknowles committed:
[BEAM-6614] Build release candidate in parallel

kennknowles committed:
[BEAM-6547] Use virtualenv for release pydoc

mxm committed:
Merge pull request #7748: [BEAM-6608] Prepare to-be-staged files early enough in Flink Runner

kennknowles committed:
[BEAM-6192] Force push to beam-website generated docs PR branch

TheNeuralBit committed:
OutputReceiver is not an annotation

dustin12 committed:
[6571] Windmill service flag (#7693)

* BEAM[6571] Adds flags for streaming engine

kennknowles committed:
mkdir -p everywhere so release script is OK with prior attempts

kennknowles committed:
svn add --force everywhere to be OK with prior attempts

kennknowles committed:
Use canonical ssh access to beam-site repo

pabloem committed:
Renaming gcp_kms_key

kennknowles committed:
Merge pull request #7735: [BEAM-6595] Improvements for release automation

adude3141 committed:
add plugin to check if build runs on jenkins

markflyhigh committed:
Fix Python logging in setup.py

lostluck committed:
Add string method to unit status

adude3141 committed:
upgrade to gogradle plugin 0.11.2

aaltay committed:
Merge pull request #7760 from lostluck/statusStrings

Add string method to unit status

tvalentyn committed:
Use BeamDeprecationWarning in BQ client code.

youngoli committed:
[BEAM-5303] Adding Python VR tests for Java Reference Runner.

Adding a gradle target and a test class for running ValidatesRunner
tests with the Java Reference Runner. Also had to add a dependency to
base_image_requirements.txt to get it to run properly. Note that this
was also causing an error when running Flink's VR tests with Docker.

chamikaramj committed:
Merge pull request #7682: [BEAM-5959] Add GCS KMS support

pabloem committed:
Reducing test size to simplify running it locally.

aaltay committed:
Merge pull request #7754 from udim/docs

Website: Update built-in IO table

aaltay committed:
Merge pull request #7712 from markflyhigh/py3-inventory

[INFRA-17335] Check more Python versions in inventory job

pabloem committed:
Merge pull request #7757 from apache/pabloem-patch-2

[BEAM-5959] Adding the dataflow_kms_key option to Python SDK

TheNeuralBit committed:
Update README.md

pabloem committed:
Merge pull request #7766 from apache/pabloem-patch-3

[BEAM-6625] Reducing test size to simplify running it locally.

aromanenko-dev committed:
Merge pull request #7701: [BEAM-6426] Fix regexp to enforce ban of non-vendored Guava

iemejia committed:
[BEAM-6426] Fix unvendored guava imports

aromanenko-dev committed:
Merge pull request #7771: [BEAM-6426] Fix unvendored guava imports

mxm committed:
Merge pull request #7763: Add plugin to check if build runs on jenkins

This allows to publish FindBugs reports in XML format on Jenkins and in HTML format on other environments.

mareksimunek committed:
[BEAM-5392] small CR corrections

Ardagan committed:
Update .test-infra/dockerized-jenkins/README.md

Co-Authored-By: TheNeuralBit <hulettbh@gmail.com>

Commit:
Update IO support matrix for Python/Streaming

swegner committed:
Merge pull request #7755: [BEAM-6615] OutputReceiver is not an annotation

swegner committed:
Merge pull request #7769: Minor tweaks to Dockerized Jenkins README

swegner committed:
Merge pull request #7768: [BEAM-5303] Adding Python VR tests for Java Reference Runner.

aaltay committed:
Merge pull request #7752 from markflyhigh/fix-disable-nocapture

[BEAM-6603] Fix Python logging in setup.py

aaltay committed:
Merge pull request #7758 from Ardagan/DatastoreIODocUpdate

State that Datastore IO is not supported in Streaming scenario.

charlesccychen committed:
Label Dataflow jobs with SDK name including Python major / minor version

mpdn committed:
Merge pull request #7770 : [BEAM-6607] Fix bug where RowCoder fails if array/map has null values

* Add test for encoding array/map with nulls

* Fix bug where null values in Row arrays and maps could not be encoded

* Move RowCoderTests to same location as other coder tests

* Deprecate array and map schema functions with nullable parameters

udim committed:
[BEAM-5959] KMS support for BigQuery - build.gradle only

aaltay committed:
Merge pull request #7779 from udim/cmek-bq-gradle

[BEAM-5959] KMS support for BigQuery - build.gradle only

markflyhigh committed:
[BEAM-6619] Add PostCommit sutie for Py3 ITs on DataflowRunner (#7756)

* [BEAM-6619] Add PostCommit sutie for Py3 ITs on DataflowRunner

udim committed:
[BEAM-5959] KMS support for BigQuery - build.gradle only - fix

aaltay committed:
Merge pull request #7780 from udim/cmek-bq-gradle

[BEAM-5959] KMS support for BigQuery - build.gradle only - fix

adude3141 committed:
[BEAM-4102] use plugins DSL instead of manual classpath

markflyhigh committed:
[BEAM-6621] Basic Bigquery IO tests on Python 3 PostCommit

swegner committed:
Merge pull request #7762: upgrade to gogradle plugin 0.11.2

aaltay committed:
Merge pull request #7778 from charlesccychen/python-version-tag

Label Dataflow jobs with SDK name including Python major / minor version

aaltay committed:
Merge pull request #7782 from markflyhigh/py3-bq-it

[BEAM-6621] Basic Bigquery IO tests on Python 3 PostCommit

robertwb committed:
Merge pull request #7730 [BEAM-6587] Make StringUtf8Coder a standard coder in Python.

iemejia committed:
Merge pull request #7690: [BEAM-4783] Fix issues created in #6181

iemejia committed:
[BEAM-6631] Update commons-compress to version 1.18

iemejia committed:
Merge pull request #7416: [BEAM-6365] Add ZStandard compression support for Java SDK

iemejia committed:
[BEAM-5987] Remove @Nullable annotation from SideInputReader.get()

Generic types are de-facto @Nullable so this annotation is redundant and
probably only existed to make FindBugs happy.

iemejia committed:
Merge pull request #7091: [BEAM-5987] Cache and share materialized side inputs between Spark tasks

LesTR committed:
BEAM-6632: fix integer overflow for interval >= 25

iemejia committed:
[BEAM-6426] Use vendored guava class for kryo registration in Spark runner

chamikaramj committed:
Merge pull request #7747: [BEAM-4359] Check for null values when encoding Key columns.

tvalentyn committed:
Make pydocs happy.

udim committed:
[BEAM-6636] Remove --kmsKey usage

pabloem committed:
Merge pull request #7765 from tvalentyn/fix_warnings

[BEAM-5667] Use BeamDeprecationWarning in BQ client code.

aaltay committed:
Merge pull request #7790 from udim/cmek-bq-gradle

[BEAM-6636] Remove --kmsKey usage

tvalentyn committed:
Restrict pyvcf dependency to Python 2 and link Jira issues tracking Py3 limitations. (#7773)

* Restrict pyvcf dependency to Python 2 and link Jira issues tracking Py3 limitations.
* Skip all VCF IO tests so that we don't need to install pyvcf on Python 3.
* Don't import VCF on Python 2.

iemejia committed:
Revert "[BEAM-5987] Remove @Nullable annotation from SideInputReader.get()"

This reverts commit 72627eb9f0b556574c23ceea06b3547d96c4107d.

melap committed:
Merge pull request #7579: Update python-pipeline-dependencies.md

iemejia committed:
Merge pull request #7789: [BEAM-6426] Use vendored guava class for kryo registration in Spark runner

charlesccychen committed:
Merge pull request #7744 from udim/cmek-bq

[BEAM-5959] KMS support for BigQuery

ihji committed:
[BEAM-6488] Portable Flink runner support for running cross-language transforms (#7709)

Multi-language support in DefaultJobBundleFactory

LesTR committed:
BEAM-6632: Use UTC timezone in test

LesTR committed:
BEAM-6632: added todo comment related to CALCITE-2837

kennknowles committed:
Merge pull request #7781: [BEAM-4102] use plugins DSL instead of manual classpath

xumingmin committed:
Merge pull request #7785: [BEAM-6632] Fixes integer overflow for interval >= 25 DAY

[BEAM-6632] Fixes integer overflow for interval >= 25 DAY

mareksimunek committed:
[BEAM-5392] GroupNonMergingWindowsFunctions.groupByKeyAndWindow description
